docs/high/file.rst

```diff
@@ -252,44 +252,46 @@ Filenames on different systems
 
 Different operating systems (and different file systems) store filenames with
 different encodings. Additionally, in Python there are at least two different
-representations of filenames, as encoded bytes (via str on Python 2, bytes on
-Python 3) or as a unicode string (via unicode on Python 2 and str on Python 3).
-The safest bet when creating a new file is to use unicode strings on all
-systems.
+representations of filenames, as encoded ``bytes`` or as a Unicode string
+(``str`` on Python 3).
+
+h5py's high-level interfaces always return filenames as ``str``, e.g.
+:attr:`File.filename`. h5py accepts filenames as either ``str`` or ``bytes``.
+In most cases, using Unicode (``str``) paths is preferred, but there are some
+caveats.
+
+.. note::
+
+   HDF5 handles filenames as bytes (C ``char *``), and the h5py :doc:`lowlevel`
+   matches this.
 
 macOS (OSX)
 ...........
 macOS is the simplest system to deal with, it only accepts UTF-8, so using
-unicode paths will just work (and should be preferred).
+Unicode paths will just work (and should be preferred).
 
 Linux (and non-macOS Unix)
 ..........................
-Unix-like systems use locale settings to determine the correct encoding to use.
-These are set via a number of different environment variables, of which ``LANG``
-and ``LC_ALL`` are the ones of most interest. Of special interest is the ``C``
-locale, which Python will interpret as only allowing ASCII, meaning unicode
-paths should be pre-encoded. This will likely change in Python 3.7 with
-https://www.python.org/dev/peps/pep-0538/, but this will likely be backported by
-distributions to earlier versions.
+Filenames on Unix-like systems are natively bytes. By convention, the locale
+encoding is used to convert to and from unicode; on most modern systems this
+will be UTF-8 by default (especially since Python 3.7, with :pep:`538`).
 
-To summarise, use unicode strings where possible, but be aware that sometimes
-using encoded bytes may be necessary to read incorrectly encoded filenames.
+Passing Unicode paths will mostly work, and Unicode paths from system
+functions like ``os.listdir()`` should always work. But if there are filenames
+that aren't in the expected encoding (e.g. on a network filesystem or a
+removable drive, or because something is misconfigured), you may want to handle
+them as bytes.
```

Comment:
I reworded this line a few times. You never* *need* to use `bytes`, because you can represent any bytes as surrogateescape in Unicode strings, and `os.fsencode()` will do the right thing. But I don't want to get into explaining that here, so I've gone with 'may want to'.

(* OK, almost never. If your locale encoding is not ASCII compatible, it doesn't hold. But that's extremely weird, so anyone affected probably either knows how to work around it or is having a very bad day trying to use Python 3 at all.)